### PR TITLE
Uploading Class work with own User creation

### PR DESCRIPTION
**Pull request by MaratulyTemirbolat**

### CODE REVIEW

File: apps/auths/admin.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from lib2to3.pgen2.token import OP
+from typing import Optional
+
+from django.contrib import admin
+from django.http import HttpRequest
+
+from auths.models import (
+    CustomUser,
+)
+
```
Comment (inmate37) [inline apps/auths/admin.py:15]: pep8

File: apps/auths/admin.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from lib2to3.pgen2.token import OP
```
Comment (inmate37) [inline apps/auths/admin.py]: imported but unused

File: apps/auths/models.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            password,
+            kwargs
+        )
+        
+        return user
+    
+
+class CustomUser(AbstractBaseUser,PermissionsMixin):
+    email = models.EmailField('Почта/Логин', unique=True)
+    is_root = models.BooleanField(default=False)
```
Comment (inmate37) [inline apps/auths/models.py]: not needed since `is_superuser`

File: apps/auths/models.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        verbose_name_plural = 'Кастомные пользователи'
+        verbose_name = 'Кастомный пользователь'
+        ordering = (
+            'email',
+        )
+    
+    def __str__(self) -> str:
+        return f'Кастомный пользователь: {self.email}'
+    
+    def save(self,*args,**kwargs) -> None:
```
Comment (inmate37) [inline apps/auths/models.py]: args, kwargs annotate

File: apps/auths/models.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        user: 'CustomUser' = self.create_user(
+            email,
+            password,
+            kwargs
+        )
+        
+        return user
+    
+
+class CustomUser(AbstractBaseUser,PermissionsMixin):
```
Comment (inmate37) [inline apps/auths/models.py]: space after comma

File: apps/auths/models.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    
+    def save(self,*args,**kwargs) -> None:
+        if self.email != self.email.lower():
+            raise ValidationError(
+                'Ваш email "%(email)s" должен быть в нижнем регистре',
+                code = 'lower_case_email_error',
+                params={'email': self.email}
+            )
+        super().save(*args,**kwargs)
+
```
Comment (inmate37) [inline apps/auths/models.py]: pep8 empty line req

File: apps/auths/models.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        user.save()
+        return user
+    
+    def create_superuser(
+        self,
+        email: str,
+        password: str,
+        **kwargs: dict
+    ) -> 'CustomUser':
+        kwargs.setdefault('is_staff',True)
```
Comment (inmate37) [inline apps/auths/models.py]: space after comma

File: apps/auths/admin.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+            return self.readonly_fields + (
+                'email',
+                'is_root',
+                'is_staff',
+                'datetime_joined',
+            )
+        return self.readonly_fields
+    
+
+admin.site.register(CustomUser,CustomUserAdmin)
```
Comment (inmate37) [inline apps/auths/admin.py]: pep8 new line req

File: apps/auths/admin.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        obj: Optional[CustomUser] = None
+        ) -> tuple:
+        if obj:
+            return self.readonly_fields + (
+                'email',
+                'is_root',
+                'is_staff',
+                'datetime_joined',
+            )
+        return self.readonly_fields
```
Comment (inmate37) [inline apps/auths/admin.py:62]:
```
if not obj:
    return self.readonly_fields

return self.readonly_fields + (
    'email',
    'is_root',
    'is_staff',
    'datetime_joined',
)

```
reason: Guard Clause

File: apps/auths/models.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            raise ValueError("Superuser must have is_staff=True")
+        if kwargs.get('is_root') is not True:
+            raise ValueError("Superuser must have is_root=True")
+
+        user: 'CustomUser' = self.create_user(
+            email,
+            password,
+            kwargs
+        )
+
```
Comment (inmate37) [inline apps/auths/models.py]: is_superuser=True req

File: apps/auths/models.py (hunk 4 above)
Comment (inmate37) [reply apps/auths/models.py]: space after commas

File: apps/auths/models.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        return f'Кастомный пользователь: {self.email}'
+    
+    def save(self,*args,**kwargs) -> None:
+        if self.email != self.email.lower():
+            raise ValidationError(
+                'Ваш email "%(email)s" должен быть в нижнем регистре',
+                code = 'lower_case_email_error',
+                params={'email': self.email}
+            )
+        super().save(*args,**kwargs)
```
Comment (inmate37) [inline apps/auths/models.py]: space after comma

File: apps/auths/models.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        )
+    
+    def __str__(self) -> str:
+        return f'Кастомный пользователь: {self.email}'
+    
+    def save(self,*args,**kwargs) -> None:
+        if self.email != self.email.lower():
+            raise ValidationError(
+                'Ваш email "%(email)s" должен быть в нижнем регистре',
+                code = 'lower_case_email_error',
```
Comment (inmate37) [inline apps/auths/models.py]: space not needed

File: apps/auths/admin.py (hunk 8 above)
Comment (inmate37) [reply apps/auths/admin.py]: space after comma

File: apps/my_new_app/admin.py [hunk 13]
```diff
@@ -84,14 +85,50 @@ class ProfessorAdmin(admin.ModelAdmin):
...
+class StudentHomeworkAdmin(admin.ModelAdmin):
+    readonly_fields = (
+        'datetime_created',
+        'datetime_updated',
+        'datetime_deleted',
+    )
+    list_filter = ('datetime_deleted',)
+    list_display = ('title', 'subject', 'get_logo', 'student')
+
+    def get_logo(self, obj: StudentHomework):
```
Comment (inmate37) [inline apps/my_new_app/admin.py]: annotate return type

File: apps/my_new_app/models.py [hunk 14]
```diff
@@ -125,38 +138,113 @@ class Professor(DateTimeCustom):
...
+        (TOPIC_RUBY, 'Ruby'),
+        (TOPIC_DELPHI, 'Delphi')
+    )
+
     full_name = models.CharField(
         verbose_name='Полное имя',
         max_length=FULL_NAME_MAX_LENGTH
     )
-        
+    account = models.OneToOneField(
```
Comment (inmate37) [inline apps/my_new_app/models.py]: user = models ...

File: apps/my_new_app/models.py [hunk 15]
```diff
@@ -125,38 +138,113 @@ class Professor(DateTimeCustom):
...
+        super().save(*args, **kwargs)
+
     def __str__(self) -> str:
         return f'Teacher: {self.full_name}, teaches: {self.topic}'
+
+
+class StudentHomeworkQuerySet(QuerySet):
+
+    def get_non_deleted(self) -> QuerySet:
+        return self.exclude(datetime_deleted__isnull=False)
```
Comment (inmate37) [inline apps/my_new_app/models.py]: if only one state its preferable to use filter rather than exclude

File: apps/my_new_app/models.py [hunk 16]
```diff
@@ -125,38 +138,113 @@ class Professor(DateTimeCustom):
...
+        verbose_name_plural = 'Домашки студентов'
+        ordering = ('title',)
+
+    def __str__(self) -> str:
+        return f'Домашка: {self.title}, выполнено: {self.student}'
+
+    def delete(self) -> None:
+        self.datetime_deleted = datetime.now()
+        self.save(
+            update_fields=['datetime_deleted']
```
Comment (inmate37) [inline apps/my_new_app/models.py:201]: image

File: apps/my_new_app/models.py [hunk 17]
```diff
@@ -125,38 +138,113 @@ class Professor(DateTimeCustom):
...
+        max_length=FILE_MAX_LENGHT_NAME,
+        verbose_name='Имя файла'
+    )
+    file = models.FileField(
+        upload_to='homeworks/files/%Y/%m/%d',
+        verbose_name='Файл'
+    )
+    homework = models.ForeignKey(
+        StudentHomework,
+        on_delete=models.CASCADE,
```
Comment (inmate37) [inline apps/my_new_app/models.py]: should we delete Homework when deleting File ?!

File: settings/base.py [hunk 18]
```diff
@@ -9,16 +9,19 @@
 For the full list of settings and their values, see
 https://docs.djangoproject.com/en/3.2/ref/settings/
 """
-import os #
-import sys #
+import os  #
+import sys  #
```
Comment (inmate37) [inline settings/base.py]: empty comment purpose ?

**Review by inmate37 — APPROVED** [review]

**Comment by inmate37** [comment]
@MaratulyTemirbolat  Ready to merge after upcoming CR fixes

File: apps/my_new_app/models.py (hunk 16 above)
Comment (MaratulyTemirbolat) [reply apps/my_new_app/models.py:201]: Обожаю вас Мади! 
АХАХАХАХАХАХАХАХАХАХАХАХАХАХ

**Pull request merged**